### DISCUSSION TITLE
How to use a custom bridge module on the Standart Host???

### DISCUSSION THREAD

**Discussion by GIANNIS-AGGELIS** [inet-framework/inet#959] [General]
Hi to all, I have seen the examples on Tsn Networks for [Asynchronous Shapers](https://inet.omnetpp.org/docs/showcases/tsn/trafficshaping/asynchronousshaper/doc/index.html).  In the Bridge layer, the module uses the stream identifier module to identify the flows by the destination port and the stream encoder module to assign PCP numbers according to the stream. 
With that in mind, I want to use something similar to a Standart Host.

I use a very simple network which comprises two Standart Hosts and one Router. Below is my network setup:
```
import inet.node.ethernet.Eth10G;
import inet.node.inet.StandardHost;
import inet.networklayer.configurator.ipv4.Ipv4NetworkConfigurator;
import inet.node.inet.Router;
network Hosts
{
    parameters:
        @display("bgb=560,381");
    submodules:
        Source: StandardHost {
            @display("p=33,98");
        }
        Destinations: StandardHost {
            @display("p=294,98");
        }
        router: Router {
            @display("p=165,98");
        }
        configurator: Ipv4NetworkConfigurator {
            @display("p=459,26");
        }
    connections:
        Source.pppg++ <--> Eth10G <--> router.pppg++;
        router.pppg++ <--> Eth10G <--> Destinations.pppg++;

}
```
Using as reference the [Bridge layer](https://doc.omnetpp.org/inet/api-current/neddoc/inet.linklayer.ethernet.common.BridgingLayer.html) from INET, I assemble a more simple module.  I copied and pasted the parameters from the Bridge layer (maybe it is not necessary). Below is the code:

```
import inet.linklayer.ethernet.common.BridgingLayer;
import inet.protocolelement.contract.IProtocolLayer;
import inet.protocolelement.processing.IProcessingDelayLayer;

module MyBridgeLayerV3 like IProtocolLayer
{
 	parameters:
        string macTableModule = default("");
        string interfaceTableModule; // relative module path of the interface table
        *.macTableModule = default(absPath(this.macTableModule));
        *.interfaceTableModule = default(this.interfaceTableModule);
        @display("i=block/layer");
    gates:
        input upperLayerIn;
        output upperLayerOut;
        input lowerLayerIn;
        output lowerLayerOut;
    submodules:
 
        streamIdentifier: <default("StreamIdentifierLayer")> like IProtocolLayer {
            @display("p=300,450");
        }
        streamCoder: <default("StreamCoderLayer")> like IProtocolLayer {
            @display("p=300,900");
        }
        
    connections:
        upperLayerIn --> { @display("m=n"); } -->streamIdentifier.upperLayerIn;
        streamIdentifier.lowerLayerOut --> streamCoder.upperLayerIn;
        streamCoder.lowerLayerOut --> { @display("m=s"); } --> lowerLayerOut;


        lowerLayerIn --> { @display("m=s"); } -->streamCoder.lowerLayerIn;
        streamCoder.upperLayerOut --> streamIdentifier.lowerLayerIn;
        streamIdentifier.upperLayerOut --> { @display("m=n"); } --> upperLayerOut;
    
}
```

If I understand right I can use my version of the bridge layer from the .ini file just by using my model through the .typename command.

```
[Config UDPAppsTestAddBrige]
network = Hosts
sim-time-limit =10s
# Configurator settings
*.configurator.dumpAddresses = true
*.configurator.dumpTopology = true
*.configurator.dumpLinks = true
*.configurator.dumpRoutes = true

# Routing settings
*.*.ipv4.arp.typename = "GlobalArp"
*.*.ipv4.routingTable.netmaskRoutes = ""

*.Source.numApps = 1
*.Source.app[0].typename = "UdpBasicApp"
*.Source.app[0].destAddresses = "Destinations"
*.Source.app[0].destPort = 5000
*.Source.app[0].messageLength = 1000B
*.Source.app[0].sendInterval = 100ms
*.Source.app[0].packetName = "UDPDataBig"

*.Destinations.numApps = 1
*.Destinations.app[0].typename = "UdpSink"
*.Destinations.app[0].localPort = 5000

*.Source.bridging.typename = "MyBridgeLayerV3"
```

When I run the simulation the network initializes but when the first packet passes the IP module the MessageDispatcher gives me the following error.

![Screenshot 2024-03-12 191059](https://github.com/omnetpp/omnetpp/assets/22667836/cddeb205-0693-46af-a4a2-607b1b04ea29)

Maybe I am missing something or I haven't configured properly my network. Does anyone know what I need to change??

**Comment by Abh4git** [comment]
I am also having same problem in a compound module I am trying to create. 
MessageDispatcher in a specific context (my case ipv4) doesnt get interfaceIdToGateIndex filled with expected interface id (101)
![image](https://github.com/inet-framework/inet/assets/5351576/3466fba8-d4cc-4bbb-aee2-0798e3f3cd1f)

In my case I am adding a QTag module in between which  is causing the issue. Somehow because of this intermediate module (maybe this bridge in the case here) the upper layer does not see the interface.  If you have direct connection (from mp to down), then it should detect the interface. I checked documentation but not sure why this happens. @rhornig any pointers on why this is happening is very useful.

**Comment by levy** [comment]
The error message is raised because the MessageDispatcher registration tables are not filled in with the necessary information. I think the root cause is something else though. The packet is supposed to go to the link layer protocol first, so it should have a DispatchProtocolReq attached. The MessageDispatcher at that level is expected to dispatch the packet based on the DispatchProtocolReq and not based on the InterfaceReq.

Is there a DispatchProtocolReq attached to the packet when it arrives at the MessageDispatcher? If so, what is the designated protocol?

**Reply by Abh4git** [reply]
Thank you for the response Levy. Its for me ipv4 protocol (UDP) coming over to the layer
![image](https://github.com/inet-framework/inet/assets/5351576/3e31399d-0c16-41c5-b039-c7458745abd5)
If I have the nl ->li direct connection, then the error is not raised, but then it wont add QTag which is actually my intention.
But if I dont have that direct connection, then eventhough from nl->InsertQTag->li link exists, it throws the error.  Any thoughts?
Am I missing something?

**Comment by levy** [comment]
Yes, that's because your insertQtag submodule is not transparent to interface registration. See the IInterfaceRegistrationListener.h header file. So the registration calls from the network interfaces are not reported towards insertQtag from li, but they are not reported to nl from insertQtag.

Unfortunately, the IEEE 802.11 MAC protocol encapsulation module resides in the network interface, so the packet must be delivered to the desired network interface first. You can try to implement that C++ interface and forward the interface registrations towards nl so that it can learn about the network interfaces and be able to dispatch packets towards them.

**Reply by Abh4git** [reply]
Thank you , yes now it is clear to me that  insertQtag is not transparent to interface registration. So in order to resolve this what shall I implement? Can you please share some more pointers? I realize there is much more to inet/omnetpp that I need to understand.
1. IInterfaceRegistrationListener.h  -> implement a C++ file for that
2. How can I have it integrated with inet to get it working
Thank you for the insights!

**Reply by levy** [reply]
The module class of insertQtag should implement the IInterfaceRegistrationListener C++ interface and forward registration calls. When you do this the MessageDispatcher will learn to forward packets towards this module.

**Reply by Abh4git** [reply]
Thank you very much!

**Reply by Abh4git** [reply]
Thanks a lot for the inputs. Still to get there. I tried out like this. But this throws interface already registered error.
void InsertQtag::handleRegisterInterface(const NetworkInterface& interface, cGate *out, cGate *in)
{

      auto outPathEnd = out->getPathEndGate();
      auto endgatefullName=outPathEnd->getOwnerModule()->getFullName();
      auto inPathStart = in->getPathStartGate();
      IInterfaceRegistrationListener *interfaceRegistration = dynamic_cast<IInterfaceRegistrationListener *>(outPathEnd->getOwner());
      if (interfaceRegistration != nullptr)
          interfaceRegistration->handleRegisterInterface(interface, inPathStart, outPathEnd);
}
 Any hints where I am I getting it wrong?

**Reply by Abh4git** [reply]
Hi @levy Looks like the inPathStart and outPathEnd need to be somehow aligned to nl and li if I get it right? Am I making sense? Thank you very much for the inputs. Still lacking some knowledge to get it going.

**Reply by Abh4git** [reply]
Thanks for the inputs. I could integrate the encoding directly on wlan to get it going. Anyway good to know how InterfaceRegistrationListener works. Maybe this is rarely used, but good to have a simpler documentation with an example for this part. Thanks Again for making me think in the right directions. Thanks, Regards, Abhilash
